runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamPool.java

```diff
@@ -0,0 +1,178 @@
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * Pool of homogeneous streams to Windmill.
+ *
+ * <p>The pool holds a fixed total number of streams, and keeps each stream open for a specified
+ * time to allow for better load-balancing.
+ */
+@ThreadSafe
+public class WindmillStreamPool<StreamT extends WindmillServerStub.WindmillStream> {
+
+  private final Duration streamTimeout;
+  private final Supplier<StreamT> streamSupplier;
+  private final List<WindmillStreamPool.StreamData<StreamT>> streams;
+  private final Map<StreamT, StreamData<StreamT>> holds;
+
+  private WindmillStreamPool(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<WindmillStreamPool.StreamData<StreamT>> streams,
+      ConcurrentHashMap<StreamT, StreamData<StreamT>> holds) {
+    this.streams = streams;
+    this.streamTimeout = streamTimeout;
+    this.streamSupplier = streamSupplier;
+    this.holds = holds;
+  }
+
+  public static <StreamT extends WindmillServerStub.WindmillStream>
+      WindmillStreamPool<StreamT> create(
+          int numStreams, Duration streamTimeout, Supplier<StreamT> streamSupplier) {
+    return new WindmillStreamPool<>(
+        streamTimeout, streamSupplier, emptyStreamList(numStreams), new ConcurrentHashMap<>());
+  }
+
+  /**
+   * Return a list of the given capacity populated with null. This list is used to load balance
+   * {@link WindmillStreamPool#getStream()} requests. It never grows in size and must be filled to
+   * capacity since the random used to grab a random stream is bound by the size of the list.
+   */
+  @SuppressWarnings("nullness")
+  private static <StreamT extends WindmillServerStub.WindmillStream>
+      List<WindmillStreamPool.StreamData<StreamT>> emptyStreamList(int capacity) {
+    List<WindmillStreamPool.StreamData<StreamT>> streams = new ArrayList<>(capacity);
+    for (int i = 0; i < capacity; i++) {
+      streams.add(null);
+    }
+
+    return streams;
+  }
+
+  @VisibleForTesting
+  static <StreamT extends WindmillServerStub.WindmillStream> WindmillStreamPool<StreamT> forTesting(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<WindmillStreamPool.StreamData<StreamT>> streams,
+      ConcurrentHashMap<StreamT, StreamData<StreamT>> holds) {
+    return new WindmillStreamPool<>(streamTimeout, streamSupplier, streams, holds);
+  }
+
+  /**
+   * Returns a stream for use that may be cached from a previous call. Each call of getStream must
+   * be matched with a call of releaseStream. If the stream has been cached but has timed out and no
+   * longer has any holds, the stream will be closed.
+   *
+   * @implNote Streams are stored in an indexed list where the index is the key to fetch the stream.
+   *     This is so that we can "randomly" fetch a stream for better load balancing.
+   */
+  public StreamT getStream() {
+    int index = ThreadLocalRandom.current().nextInt(streams.size());
```

Comment:
since size is 1 generally how about special casing that to avoid random?

Reply:
done.